### PR TITLE
PROVCAU-141 Custom idno handling for explodeValue() and getSortableValue().

### PR DESCRIPTION
**Pull request by bri-gaia**

### CODE REVIEW

File: app/lib/IDNumbering/ProvMultipartIDNumber.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	/**
+	 * Returns sortable value padding according to the custom PROV rules
+	 *
+	 * @param string $value Value from which to derive the sortable value. If omitted the current value is used. [Default is null]
+	 * @return string The sortable value
+	 */
+	public function getSortableValue($value=null) {
+		preg_match_all('/([a-zA-Z]+|\d+)/', $value, $matches);
+		$output = array_map(function($v) {
+			$padding = 20;
```
Comment (bri-gaia) [inline app/lib/IDNumbering/ProvMultipartIDNumber.php:43]: I was wrong @kehh this doesn't seem to pull from the config at all in the parent. Probably could be retrofitted if we have more time.

File: app/lib/IDNumbering/ProvMultipartIDNumber.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			$parent_value = $this->getParentValue();
+			if (!empty($parent_value)) {
+				$v_proc = preg_replace( "!^" . preg_quote( $parent_value, '!' ) . "!", "_PARENT_", $value );
+
+				if (str_contains( $v_proc, '_PARENT_' )) {
+					$element_vals = $this->explodeValue( $v_proc, true );
+					$element_vals = array_map( function ( $v ) use ($parent_value) {
+						return preg_replace( "!^_PARENT_!", $parent_value, $v );
+					}, $element_vals );
+					return $element_vals;
```
Comment (bri-gaia) [inline app/lib/IDNumbering/ProvMultipartIDNumber.php:26]: This works when the ID numbers are "correct" but gets super weird, really quick if the values don't match the config returned from $this->getElementInfo()

File: app/lib/IDNumbering/ProvMultipartIDNumber.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	# -------------------------------------------------------
+	/**
+	 * Initialize the plugin
+	 *
+	 * @param string $format A format to set as current [Default is null]
+	 * @param mixed $type A type to set a current [Default is __default__]
+	 * @param string $value A value to set as current [Default is null]
+	 * @param Db $db A database connection to use for all queries. If omitted a new connection (may be pooled) is allocated. [Default is null]
+	 */
+	public function __construct($format=null, $type=null, $value=null, $db=null) {
```
Comment (kehh) [inline app/lib/IDNumbering/ProvMultipartIDNumber.php]: strip this out plz

File: app/lib/IDNumbering/ProvMultipartIDNumber.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	public function explodeValue($value, $parentPlaceholder=false) {
+		$separator = $this->getSeparator();
+
+		if ($value && !$parentPlaceholder && !$separator && $this->formatHas('PARENT', 0)) {
+			// starts with PARENT element so replace with placeholder as parent value may include separators
+			$parent_value = $this->getParentValue();
+			if (!empty($parent_value)) {
+				$v_proc = preg_replace( "!^" . preg_quote( $parent_value, '!' ) . "!", "_PARENT_", $value );
+
+				if (str_contains( $v_proc, '_PARENT_' )) {
```
Comment (kehh) [inline app/lib/IDNumbering/ProvMultipartIDNumber.php:21]: throw an exception

Comment (kehh) [reply app/lib/IDNumbering/ProvMultipartIDNumber.php:21]: also think about adding a custom config for consignments:

```
		consignments = {
			separator =,
			
			elements = {
				series_identifier = {
					type = PARENT,
					editable = 0,
					description = _(Series identifier),
					separator_pattern = '/'
				},
```

File: app/lib/IDNumbering/ProvMultipartIDNumber.php (hunk 1 above)
Comment (kehh) [reply app/lib/IDNumbering/ProvMultipartIDNumber.php:43]: don't worry about pulling from config as this is custom code so we can change the code as easily as changing the config

File: app/lib/IDNumbering/ProvMultipartIDNumber.php (hunk 4 above)
Comment (bri-gaia) [reply app/lib/IDNumbering/ProvMultipartIDNumber.php:21]: @kehh I'm currently not wrapping my head around these. Going to put a pin in this and work on the other MR.

Comment (bri-gaia) [reply app/lib/IDNumbering/ProvMultipartIDNumber.php:21]: @kehh @natasha-gaia throwing an exception here is very disruptive as explodeValue is called in a _lot_ of places, including read contexts. 

I'm instead adding a check to my migration to ensure the generated IDNO is somewhat reasonable. Happy to revisit this for more discussion.

Comment (natasha-gaia) [reply app/lib/IDNumbering/ProvMultipartIDNumber.php:21]: I feel thats a fair call to do the check in the migration directly

**Comment by natasha-gaia** [comment]
LGTM

**Review by natasha-gaia — APPROVED** [review]
LGTM

**Pull request merged**